Starting the document on TLA+ idioms (#325)
informalsystems/apalache: docs/idiomatic/000keep-minimum-state-variables.md

# Idiom 0: Keep state variables to the minimum

## Description

_A good TLA+ specification minimizes the computation state and makes it visible_.

> 
> TLA+ does not have special syntax for assignments to the variables.  For a good

Reviewer: ```suggestion

TLA+ does not have special syntax for variable assignment.  For a good
```

Reviewer: BTW, this statement is controversial. TLA+ does have variable assignment, which is what idiom 001 states explicitly:

> TLA+ uses the same glyph, `=`, for assignment, asserting equality, and binding variables.

Perhaps we can rephrase the whole sentence to be consistent with idiom 001.

Reviewer 2: Perhaps something like this?

```suggestion

TLA+ does not have special syntax that distinguishes mutable variable assignment from equality constraints.  For a good
```

> # Idiom 0: Keep state variables to the minimum

Author: Yes. I think this is also my intuition.

> # Idiom 0: Keep state variables to the minimum

Author: I will think about where to add it. If you have a good phrase for it, just add it.

> # Idiom 0: Keep state variables to the minimum

Reviewer 2: How about right here?

```suggestion
# Idiom 0: Keep state variables to the minimum

In imperative programming, it is common to use mutable variable assignments liberally, but to exercise caution whenever mutable variables have a global scope. In TLA+, mutable variables are always global, so it is important to use them carefully and in a way that accurately reflects the global state of the system you are specifying.  
```

  1. To make the code concise.

Point 1 is a non-issue in TLA+, as it is mostly executed in the reader's brain,
and people are probably less efficient in caching expressions than computers.
Points 2 and 3 can be nicely addressed with LET-definitions in TLA+. Hence,
there is no need for auxiliary variables.

Usually, we should minimize the specification state, that is, the scope of the data
structures that are declared with `VARIABLES`. It does not mean that one variable
is always better than two. It means that what is stored in `VARIABLES` should be
absolutely necessary to describe the computations or the observed properties.

## Advantages

By avoiding auxiliary state variables, we localize the updates to the state.
This improves specification readability. It also helps the tools, as large parts
of the specification become deterministic.

## Disadvantages

Sometimes, we have to expose the internals of the computation. For instance,
if we want to closely monitor the values of the computed expressions, when using
the specification for model-based testing.

## Example

Consider the following implementation of [Bubble sort] in Python:

```python
    my_list = [5, 4, 3, 8, 1]

>     changed = True

Author: ```suggestion
    finished = False
```

    my_list_len = len(my_list)  # cache the length

>     while changed:
>         changed = False

Author: ```suggestion
    while not finished:
        finished = True
```

        if my_list_len > 0:
            prev = my_list[0]       # save the first element to use in the loop
        for i in range(1, my_list_len):
            current = my_list[i]
            if prev <= current:
                # save current for the next iteration
                prev = current
            else:
                # swap the elements
                my_list[i - 1] = current
                my_list[i] = prev

>                 changed = True

Author: ```suggestion
                finished = False
```

```

Notice that we have introduced three local variables to optimize the code:

  - `my_list_len` to cache the length of the list,
  - `prev` to cache the previously accessed element of the list,
    in order to minimize the number of list accesses,
  - `current` to cache the iterated element of the list.

In TLA+, one usually does not introduce local variables for the intermediate
results of the computation, but rather introduces variables to represent the
essential part of the algorithm state. (While we have spent some time on code
optimization, we might have missed the fact that our sorting algorithm is not
as good as [Quicksort].) In the above example, the essential variables are

> `finished` and `my_list`.

Author: Good point. I guess, I have refactored the TLA+ spec without looking back at the python code. Updated the python code.

  1. To break down a large expression into a series of smaller expressions, 
  1. To make the code concise.

Point 1 is a non-issue in TLA+, as it is mostly executed in the reader's brain,
and people are probably less efficient in caching expressions than computers.
Points 2 and 3 can be nicely addressed with LET-definitions in TLA+. Hence,
there is no need for auxiliary variables.

Usually, we should minimize the specification state, that is, the scope of the data
structures that are declared with `VARIABLES`. It does not mean that one variable
is always better than two. It means that what is stored in `VARIABLES` should be
absolutely necessary to describe the computations or the observed properties.

## Advantages

By avoiding auxiliary state variables, we localize the updates to the state.
This improves specification readability. It also helps the tools, as large parts
of the specification become deterministic.

## Disadvantages

Sometimes, we have to expose the internals of the computation. For instance,
if we want to closely monitor the values of the computed expressions, when using
the specification for model-based testing.

> 

Author: ```suggestion

Sometimes, we have to break this idiom to make the specification more readable. Here is an example by Markus Kuppe. The specification of [BlockingQueue](https://github.com/lemmy/BlockingQueue/blob/3a66f46f6f5703f2863f71baaf0aedaaee58836f/BlockingQueueSplit.tla#L16-L51) that has one more variable is easier to read than [the original specification](https://github.com/lemmy/BlockingQueue/blob/3a66f46f6f5703f2863f71baaf0aedaaee58836f/BlockingQueue.tla) with a minimal number of variables.
```
